### PR TITLE
port `utils::mkdir_parents` to rust

### PR DESCRIPTION
**Pull request by SeanBE**
See #334. I'm casting mode_t to a signed int before sending it over. Not sure if there was a more appropriate way of going about it.

### CODE REVIEW

File: rust/libnewsboat/Cargo.toml [hunk 1]
```diff
@@ -46,3 +46,4 @@ features = [ "gettext-system" ]
 tempfile = "3"
 proptest = "0.7"
 section_testing = "0.0.4"
+tempdir = "0.3.7"
```
Comment (Minoru) [inline rust/libnewsboat/Cargo.toml]: According to [the crates.io](https://crates.io/crates/tempdir), "The tempdir crate is being merged into tempfile". We are already using `tempfile` in tests, even [here in utils](https://github.com/newsboat/newsboat/blob/c01acf47142637b8959036585bc30eaf28c30052/rust/libnewsboat/src/utils.rs#L816-L834). Please revert the changes to Cargo.lock and Cargo.toml, and use `tempfile` instead.

File: rust/libnewsboat/src/utils.rs [hunk 2]
```diff
@@ -477,6 +479,19 @@ pub fn getcwd() -> Result<PathBuf, io::Error> {
...
 }
 
+/// Recursively create directories if missing and set permissions accordingly.
+pub fn mkdir_parents(p: &str, mode: u32) -> isize {
+    let path = Path::new(p);
+    let result = fs::DirBuilder::new()
+        .mode(mode)
+        .recursive(true) // directories created with same security and permissions
+        .create(path);
+    match result {
```
Comment (Minoru) [inline rust/libnewsboat/src/utils.rs]: I think it would be better to do this conversion in `libnewsboat-ffi`. This way, our C++ code can keep working with a (somewhat uninformative) integer, and our Rust code is forced to handle the `Result` properly.

File: rust/libnewsboat/src/utils.rs [hunk 3]
```diff
@@ -477,6 +479,19 @@ pub fn getcwd() -> Result<PathBuf, io::Error> {
...
+        Ok(()) => 0,
+        Err(_) => -1,
```
Comment (Minoru) [inline rust/libnewsboat/src/utils.rs]: In C and C++, -1 indicates that there are an error, but the error itself is communicated out of band via `errno` variable—which this function doesn't set. Can you please adjust [the only place that checks `errno`](https://github.com/newsboat/newsboat/blob/c01acf47142637b8959036585bc30eaf28c30052/src/pbcontroller.cpp#L103-L110) to not use the variable? Just fail if ret is -1. It's safe not to check that errno is EEXIST because your test establishes that re-creating directories with this `make_parents` returns Ok.

File: include/rs_utils.h [hunk 4]
```diff
@@ -78,6 +78,8 @@ char* rs_program_version();
 
 unsigned int rs_newsboat_version_major();
 
+int rs_mkdir_parents(const char* path, const unsigned int mode);
```
Comment (Minoru) [inline include/rs_utils.h]: It's better to use an integer of fixed width, `std::uint32_t` in this case, because Rust expects precisely that size and signedness.

I notice we make the same mistake of using `int` all over our FFI bindings. I'll open an issue about this.

File: src/utils.cpp [hunk 5]
```diff
@@ -944,44 +943,7 @@ unsigned int utils::gentabs(const std::string& str)
...
-		// It's not an error if the directory already exists. This happens when
-		// `p` ended with a slash, in which case the loop above creates all the
-		// path components.
-		if (result == -1 && errno == EEXIST) {
-			result = 0;
-		}
-	}
-
-	return result;
+	return rs_mkdir_parents(p.c_str(), (int) mode);
```
Comment (Minoru) [inline src/utils.cpp]: `(int)mode` is a C-style cast, and it's very powerful (something like Rust's `mem::transmute`). It's better to use one of C++'s more restricted casts, in this case `static_cast<std::uint32_t>(mode)`.

I'm actually not sure if `mode_t` is always 32 bits, but upcasting doesn't hurt, and an attempt in downcasting from 64 bits or whatever will give us a compilation error—all acceptable outcomes.

File: rust/libnewsboat/src/utils.rs [hunk 6]
```diff
@@ -477,6 +479,19 @@ pub fn getcwd() -> Result<PathBuf, io::Error> {
     env::current_dir()
 }
 
+/// Recursively create directories if missing and set permissions accordingly.
+pub fn mkdir_parents(p: &str, mode: u32) -> isize {
+    let path = Path::new(p);
```
Comment (Minoru) [inline rust/libnewsboat/src/utils.rs]: I think it's better to accept `&Path` in Rust, and let the FFI bindings convert from `const char*` to that. This way, our C++ code keeps working and our Rust code is forced to use a more appropriate `Path` type.

**Comment by Minoru** [comment]
Also, src/libnewsboat/src/configpaths.rs [already has](https://github.com/newsboat/newsboat/blob/c01acf47142637b8959036585bc30eaf28c30052/rust/libnewsboat/src/configpaths.rs#L374-L380) some similar code, can you please rewrite that `try_mkdir` in terms of `mkdir_parents`, to avoid duplication?

File: rust/libnewsboat/Cargo.toml (hunk 1 above)
Comment (SeanBE) [reply rust/libnewsboat/Cargo.toml]: Ah! Reminder to myself to check crates.io next time. I went straight to the docs.

File: rust/libnewsboat/src/utils.rs (hunk 2 above)
Comment (SeanBE) [reply rust/libnewsboat/src/utils.rs]: Totally! I'm just creating more unnecessary work haha. I've moved the conversions over.

File: rust/libnewsboat/src/utils.rs (hunk 3 above)
Comment (SeanBE) [reply rust/libnewsboat/src/utils.rs]: dropped the condition and removed errno completely incl. the log statement.

File: include/rs_utils.h (hunk 4 above)
Comment (SeanBE) [reply include/rs_utils.h]: Done. I've also pushed your comments into the commit bodies.

File: src/utils.cpp (hunk 5 above)
Comment (SeanBE) [reply src/utils.cpp]: Maybe I should leave a comment for future reference?

**Comment by SeanBE** [comment]
I appreciate the feedback/comments. Learnt a ton!

File: rust/libnewsboat/src/configpaths.rs [hunk 7]
```diff
@@ -9,6 +9,8 @@ use std::os::unix::fs::DirBuilderExt;
 use std::path::{Path, PathBuf};
 use xdg;
 
+use utils;
```
Comment (Minoru) [inline rust/libnewsboat/src/configpaths.rs]: We don't split imports into sections at the moment, so please put that import before `use xdg`.

File: src/utils.cpp (hunk 5 above)
Comment (Minoru) [reply src/utils.cpp]: What would the comment say? If this code fails to compile, I think the error will be enugh, saying something like `no matching conversion for static_cast from 'mode_t' (aka 'foo bar') to 'std::uint32_t' (aka 'baz quux')`.

**Comment by Minoru** [comment]
The remaining issues are trivial, so I'm approving this. Good job! You can rewrite history now if you want, to get rid of the revert commit. I'll merge this when you've fixed the import.

As we already discussed on IRC, clang-format is not used at the moment because different versions produce different formatting. The only formatter we use is for Rust, and [your code passed it without a hitch](https://travis-ci.org/newsboat/newsboat/jobs/578579892). You don't have to get clang-format working for this PR, so don't worry about it.

**Comment by Minoru** [comment]
[`cargo fmt` is unhappy about an extra empty line](https://travis-ci.org/newsboat/newsboat/jobs/579045486#L219), which is not worth an extra review round—I'll just fix it after merging. Thanks for doing all the hard work, @SeanBE!
